### PR TITLE
feat: add `intRange` type

### LINKED ISSUE #669: Proposal: `Enumerate` and `IntRange`
Useful for when we are working with restrict numeric values

definition:
```ts
type Enumerate<N extends number, Acc extends number[] = []> = Acc['length'] extends N
  ? Acc[number]
  : Enumerate<N, [...Acc, Acc['length']]>

type IntRange<F extends number, T extends number> = Exclude<Enumerate<T>, Enumerate<F>> | T
```
So it can be used:

```ts
type ValidAmount = Enumerate<4> //  0 | 1 | 2 | 3
type WeekDay = IntRange<1, 7> // 1 | 2 | 3 | 4 | 5 | 6 | 7
```

### PR DESCRIPTION
**Pull request by Emiyaaaaa**
Fixes #669

### CODE REVIEW

**Comment by sindresorhus** [comment]
Thanks for contributing :)

**Comment by sindresorhus** [comment]
How does this compare to https://twitter.com/anuraghazru/status/1573328100246822917 and https://github.com/sindresorhus/type-fest/issues/669#issue-1865214179 ?

**Comment by sindresorhus** [comment]
I'm a little bit concerned about possible type complexity explosion here. Have you tested it with larger ranges?

Maybe it would make sense to add an overload to use the [simpler way](https://github.com/sindresorhus/type-fest/issues/669#issue-1865214179) when "step" is not specified?

**Comment by sindresorhus** [comment]
// @lucasteles @tommy-mitchell

**Comment by Emiyaaaaa** [comment]
> I'm a little bit concerned about possible type complexity explosion here. Have you tested it with larger ranges?
> 
> Maybe it would make sense to add an overload to use the [simpler way](https://github.com/sindresorhus/type-fest/issues/669#issue-1865214179) when "step" is not specified?

You are right, with "step" is significantly slower than without "step". I found `EqualOrMoreThan` is very performance-impairing, but it is necessary
 for "step"
I should use overload and try to resolve performance question with `Step`.

ps: range limit is 0-999 in any implementation

**Comment by Emiyaaaaa** [comment]
> How does this compare to https://twitter.com/anuraghazru/status/1573328100246822917 and [#669 (comment)](https://github.com/sindresorhus/type-fest/issues/669#issue-1865214179) ?

these are the same principle: generate a `Start` to `End` tuple like `type RangeArray<5, 10> = [5,6,7,8,9]`, and transform it to union.

twitter and #669 way is generate a number tuple A of "0 - Start" and anonter number tuple B of "0 - End". then `Exclude<B, A>`

My why is generate a tuple of "0 - End", and use any other type when tuple index is small than `Start`, (e.x. `type IntRangeArray<3, 5> = [never, never, never, 3, 4]`), than transform tuple to union and exclude `never`, so my way just create one tuple.

twitter's `FloatRannge`'s implementation method is transform `IntRange` result to string, and add the decimals, then transform to number union

**Comment by Emiyaaaaa** [comment]
> > I'm a little bit concerned about possible type complexity explosion here. Have you tested it with larger ranges?
> > Maybe it would make sense to add an overload to use the [simpler way](https://github.com/sindresorhus/type-fest/issues/669#issue-1865214179) when "step" is not specified?
> 
> You are right, with "step" is significantly slower than without "step". I found `EqualOrMoreThan` is very performance-impairing, but it is necessary for "step"
> 
> I should use overload and try to resolve performance question with `Step`.
> 
> ps: range limit is 0-999 in any implementation

@sindresorhus  latest commit used overload and improved performance when use `Step`

File: source/int-range.d.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+type PrivateIntRange<
+	Start extends number,
+	End extends number,
+	Step extends number,
+	Gap extends number = Subtract<Step, 1>, // The gap between each number, gap = step - 1
+	List extends unknown[] = BuildTuple<Start, never>, // The final `List` is [...StartLengthTuple, ...[number, ...GapLengthTuple], ...[number, ...GapLengthTuple], ... ...], So can initialize the `List` with [...StartLengthTuple]
+> = EqualOrMoreThan<List['length'], End> extends true
+	? Exclude<List[number], never> // All unused elements are `never`, so exclude them
+	: PrivateIntRange<Start, End, Step, Gap, [...List, List['length'], ...BuildTuple<Gap, never>]>;
+
```
[the remark this answers is not in the document]
Comment (Emiyaaaaa) [reply source/int-range.d.ts]: > 

Outdated code

**Comment by sindresorhus** [comment]
Thanks :)

**Pull request merged**